src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java

```diff
@@ -244,13 +205,46 @@ public JsonWebKey supplyKey(
 
         private final Map<String, Long> scopeBitsByName = new CopyOnWriteHashMap<>();
 
-        private long realmBit;
+        private final long realmId;
         private long nextScopeBitShift;
 
         private OAuthRealm(
-            long realmBit)
+            long realmId)
+        {
+            this.realmId = realmId;
+        }
+
+        private long resolve(
+            String[] scopeNames)
+        {
+            long authorization = realmId;
+            // if not already there, add the scope to the map, assign each scope a bit
+            // which determines which low bit will be flipped if that scope is present
+            for (int i = 0; i < scopeNames.length; i++)
+            {
+                final String scope = scopeNames[i];
+                // check if scope's bit has been set and if scope can be added
+                if(!scopeBitAssigned(scope) && !supplyScopeBit(scope))
+                {
+                    throw new IllegalStateException("Too many scopes");
+                }
```

Comment:
Is `scopeBitAssigned` no longer required?
Also, `supplyScopeBit` should return the value not a boolean, then no need to also call `getScopeBit` below.
Check for too many scopes should be in new method called `assignScopeBit`, passed as a method reference `this::assignScopeBit` to `scopeBitsByName.computeIfAbsent(...)`.

Reply:
Please apply the changes above.